docs: Firestore emulator for driver tests design (#1246)
platform-platform/monorepo: metrics/web/docs/features/firestore_emulator_for_driver_tests/01_firestore_emulator_for_driver_tests.md

1. Run the Firestore emulator.
2. Create test data in the Firestore emulator using the Firestore emulator UI. 
3. Run the `firebase emulators:export ../web/test_driver/firestore_test_data` command to export test data to the `export_directory` from inside the `metrics/firebase` directory.

The process of the test data creation will be performed only once and, the test data will be added to the version control.

Once we have exported data from the Firestore emulator, we can run an emulator with the exported data using the following command each time before running the driver tests:

```bash
firebase emulators:start --import=../web/test_driver/firestore_test_data
```

### Architecture

> Fundamental structures of the feature and context (diagram).

Consider the class diagram showing the main feature classes and relationships:

![Firestore emulator class diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/firestore_emulator_design/metrics/web/docs/features/firestore_emulator_for_driver_tests/diagrams/firestore_emulator_class_diagram.puml)

> ### Program

Reviewer: I believe it would be great to explain the process of configuring the application under tests to use the Firestore emulator instance instead of the production one. What do you think?

> Detailed solution description to class/method level.

> As we want to use the Firebase emulator to run integration tests, we should add a parameter - `use-firestore-emulator`, that determines whether tests should use the production database or run the local emulator with test data. 
> 
> To connect the application to the running Firestore emulator, we need to know the emulator's port, so we also should add a `firestore-emulator-port` option to the driver tests script.

Reviewer: ```suggestion
As we want to use the Firebase emulator to run integration tests, we should add a parameter - `use-firestore-emulator`, that determines whether tests should use the production database or run the local emulator with test data. Also, to connect the application to the running Firestore emulator, we need to know the emulator's port, so we should add a `firestore-emulator-port` option to the driver tests script. This option represents the port the `Firestore emulator` will run on and provides an ability to change it if it is required.
```

To group these parameters and simplify the logic of retrieving their values inside the integration tests we should create the `FirestoreEmulator` model.

> The `FlutterWebDriver` class, that responsible for running the application and driver tests, can use the model's values and pass them to the environment using the `dartDefine` method, so we can access these values in the integration tests using the `FirestoreEmulator.fromEnvironment()` method.

Reviewer: ```suggestion
To pass the Firestore emulator configuration to the application under tests, we are going to update the `FlutterWebDriver` class and a new `dart-define` option containing the `Map` representation of the Firestore emulator configuration. It allows us to easily access the configuration from the application under tests using the `FirestoreEmulator.fromEnvironment()` factory method.
```
What do you think?

So with that in place, we can run the emulator by using the following command:

```bash
    firebase emulators:start --only firestore --import=import_directory
```

Now the emulator is running and it is time to use it in the app's integration tests.

In the `setUpAll` method of the tests we need to configure the Firestore instance to use the emulator:

```dart
setupAll(() {
    Firestore.instance.settings(
        host: 'localhost:8080',
        sslEnabled: false,
        persistenceEnabled: false,
    );
});
```

The specified port in the `host` argument must be equal to the emulator's port.

# Design

### User Interface

> How users will interact with the feature (API, CLI, Graphical interface, etc.).

The feature is introduced by adding additional parameters to the integration tests running command.

The next table contains the definitions of the parameters with their default values.

| Parameter         | Default | Description                                                           |
| ----------------- | -------- | --------------------------------------------------------------------- |

> | **use-emulator**  | true     | Determines if the integration tests are using the `Firebase Emulator`. |

Reviewer: ```suggestion
| **use-firestore-emulator**  | true     | Determines if the integration tests are using the `Firebase Emulator`. |
```
Would it be a bit more clear?

| **emulator-port** | 8080     | Specifies the emulator's running port.                                  |

The following code snippet shows an example of using the new parameters:

```bash
dart test_driver/main.dart --use-emulator --emulator-port=8081
```

Or, if you accept the defaults, you can omit them.

As described in the [Prototyping](#prototyping) section, we should run integration tests once the `Firebase Emulator` started.

### Database

> How relevant data will be persisted.

Integration tests interacts with the application that uses data from the database. 

As we don't want to use the production environment, we should use the `Firebase local emulator`.

The emulator has no data as a default, so we need to create necessary collections.

It is not convenient to create it every time the emulator starts. So, for this purposes, we need to create all collection once and export it using the following command:

```bash
firebase emulators:export export_directory
```

_**Note:** You can place just `.` instead of the `export_directory` name and the firebase creates the directory, named `firestore_export` as a default._

> From now, we can run emulator with the exported data using the following command:

Reviewer: ```suggestion
Once we have exported data from the Firestore emulator, we can run an emulator with the exported data using the following command:
```

```bash
firebase emulators:start --import=export_directory
```

### Program

> > Detailed solution description to class/method level.

Reviewer: Should we explain the way of running the `Firestore emulator` in terms of the `test_driver` here?

Now we can run the emulator with the initial data by adding the following flag `--import=import_directory`, where the `import_directory` is the directory that holds the exported data.

So with that in place, we can run the emulator by using the following command:

```bash
    firebase emulators:start --only firestore --import=import_directory
```

Now the emulator is running and it is time to use it in the app's integration tests.

In the `setUpAll` method of the tests we need to configure the Firestore instance to use the emulator:

```dart
setupAll(() {
    Firestore.instance.settings(
        host: 'localhost:8080',
        sslEnabled: false,
        persistenceEnabled: false,
    );
});
```

The specified port in the `host` argument must be equal to the emulator's port.

# Design

### Architecture

> Fundamental structures of the feature and context (diagram).

> ![Firestore emulator class diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/firestore_emulator_design/metrics/web/docs/features/firestore_emulator_for_driver_tests/diagrams/firestore_emulator_class_diagram.puml)

Reviewer: Would it be better to move this diagram to be closer to the section where we are explaining the process of configuring new arguments and passing them to the Flutter driver tests?

### User Interface

> How users will interact with the feature (API, CLI, Graphical interface, etc.).

The feature is introduced by adding additional parameters to the integration tests running command.

The next table contains the definitions of the parameters with their default values.

| Parameter         | Default | Description                                                           |
| ----------------- | -------- | --------------------------------------------------------------------- |
| **use-firestore-emulator**  | true     | Determines if the integration tests are using the `Firebase Emulator`. |

> | **emulator-port** | 8080     | Specifies the emulator's running port.                                  |

Reviewer: ```suggestion
| **firestore-emulator-port** | 8080     | Specifies the emulator's running port.                                  |
```
?

### Database

> How relevant data will be persisted.

Integration tests interact with the application that uses data from the database. As we don't want to use the production environment, we should use the `Firebase local emulator`.

By default, the emulator starts with no data, so we should add test data used during testing. The best approach to do so is importing the test data to the Firestore emulator once running. But before exporting the test data, we should create it. The `emulators:export` command allows to export the data from the running Firestore emulator. 

So, to create test data for the Firestore emulator, we should follow the next steps: 

1. Run the Firestore emulator.
2. Create test data in the Firestore emulator using the Firestore emulator UI. 
3. Run the `firebase emulators:export export_directory` command to export test data to the `export_directory`.

_**Note:** The `export_directory` in our case should be a `firebase/tests/firestore/test_data`._

Once we have exported data from the Firestore emulator, we can run an emulator with the exported data using the following command:

```bash

> firebase emulators:start --import=export_directory

Reviewer: It would be great to have a diagram showing the main classes and relationships used to run such commands. Also, It would be useful to have a `sequence` diagram showing the process of running driver tests using the Firestore emulator.

> 3. Run the `firebase emulators:export export_directory` command to export test data to the `export_directory`.
> 
> _**Note:** The `export_directory` in our case should be a `firebase/tests/firestore/test_data`._
> 
> Once we have exported data from the Firestore emulator, we can run an emulator with the exported data using the following command:
> 
> ```bash
> firebase emulators:start --import=export_directory

Reviewer: Would it be better to place the Firestore emulator test data under the `web/test_driver/firestore_test_data` package? Also, does it make sense to mention that the process of the test data creation will be performed only once and, the test data will be added to the version control? 
```suggestion
3. Run the `firebase emulators:export firestore_test_data` command to export test data to the `export_directory`.

Once we have exported data from the Firestore emulator, we can run an emulator with the exported data using the following command each time before running the driver tests:

```bash
firebase emulators:start --import=firestore_test_data
```

```

### Program

> Detailed solution description to class/method level.

As we want to use the Firestore emulator to run integration tests, we should add a parameter - `use-firestore-emulator`, that determines whether tests should use the production database or run the local emulator with a test data. 

> Since the local emulator runs on a specific port we want to connect the application to that specific port via the `emulator-port` parameter.

Reviewer: ```suggestion
Also, to be able to configure the Firestore emulator port we are going to add an `emulator-port` option to the driver tests script.
```
